# <span style="text-decoration: underline">Assessment 2019 Fundamentals of Data Analysis</span>

## This assessment concerns the well-known tips dataset and the Python packages seaborn and jupyter.

1. Create a jupyter notebook that uses descriptive statistics and plots to describe the tips dataset.
1. To the above jupyter notebook add a section that discusses and analyses whether there is a relationship 
between the total bill and tip amount.
1. Using the same notebook, analyse the relationship between the
variables within the dataset. You are free to interpret this as you wish — for
example, you may analyse all pairs of variables, or select a subset and analyse
those.
***

# <span style="text-decoration: underline">Introduction</span>
What are tips?
A sum of money customarily given by a client or customer to certain service sector workers for the service they have performed, in addition to the basic price of the service.


According to https://www.tripadvisor.ie 
>There is not a strong tipping culture in Ireland but many locals and visitors tend to tip for certain services, 
>in particular restaurants, taxis and personal services (e.g. hairdressing). 
>If you want to tip and are in doubt about how much, think 10%. All tips are at your discretion!

However, in the US where this dataset was collected tips form a substantial amount of an employees' income due to the nominal hourly rate paid. This data can be useful to managers or the staff themselves in order to ensure that there is equity in the receipt of tips when they are not gathered and shared out pro rata.  

An important point to bear in mind for this assessment is to view the data set and subsequent analyses with fresh eyes.  Forget your assumptions about tipping culture and focus only on what can be proven with data.

[An introduction to seaborn](https://seaborn.pydata.org/introduction.html) states that the tips dataset is
>very boring but quite useful for demonstration.

So consider yourself warned!
The alleged dullness of the data is, however, more than made up for in the beautiful plots that can  be created using the seaborn library.
***


# <span style="text-decoration: underline">1. Create a jupyter notebook that uses descriptive statistics and plots to describe the tips dataset.</span>

## Descriptive Statistics

Descriptive statistics describe data. 
It’s a way to summarise and organise collected data into a manageable format that is easy to understand.<br>

In [9]:
#To begin,import the required python packages, each is given an abbreviation in order to save keystrokes.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
#Mandatory dependencies for seaborn are: NumPy, Matplotlib, Pandas
import seaborn as sns

In [10]:
#The tips dataset is built into seaborn
#View all available datasets in seaborn using this command
sns.get_dataset_names()

['anscombe',
 'attention',
 'brain_networks',
 'car_crashes',
 'diamonds',
 'dots',
 'exercise',
 'flights',
 'fmri',
 'gammas',
 'iris',
 'mpg',
 'planets',
 'tips',
 'titanic']

Even if you don't know anything about a particular dataset there are a few simple commands to produce an overview.<br>
Immediately it can be seen that 244 rows and 7 columns categories in the dataset.

In [13]:
tips = sns.load_dataset("tips")
#code verbatim from https://www.kaggle.com/ssouhassou/data-viz-a-modest-essay-on-the-dataset-tips
tips.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
total_bill    244 non-null float64
tip           244 non-null float64
sex           244 non-null category
smoker        244 non-null category
day           244 non-null category
time          244 non-null category
size          244 non-null int64
dtypes: category(4), float64(2), int64(1)
memory usage: 7.2 KB


In [ ]:
df=tips
df.shape
#putting the column label into square brackets produces output of a list from that column in data frame format

In [15]:
tips.describe()
#if parentheses ()are not used, all columns will display but no useful summary statistics!
#According to pandas 0.25.1 documentation:
#For numeric data, the result’s index will include count, mean, std, min, max as well as lower, 50 and upper percentiles. 
#By default the lower percentile is 25 and the upper percentile is 75. The 50 percentile is the same as the median.

total_bill         tip        size
count  244.000000  244.000000  244.000000
mean    19.785943    2.998279    2.569672
std      8.902412    1.383638    0.951100
min      3.070000    1.000000    1.000000
25%     13.347500    2.000000    2.000000
50%     17.795000    2.900000    2.000000
75%     24.127500    3.562500    3.000000
max     50.810000   10.000000    6.000000

In [ ]:
#The "tips" dataset is a commonly used one that is built into the seaborn Python data visualization library
tips = sns.load_dataset("tips")
#No requirement to type "print" - call it to be printed with just the name of the dataset.
tips

In [ ]:
tips.head()
#The head() command returns 5 as a default when no argument is entered

In [ ]:
tips.tail(3)
#this time 3 is passed into the argument, otherwise 5 outputs would print

Checking the "head" and "tail" entries on the dataset is a useful exercise to quickly<br>
check the integrity of the dataset, particularly if it is a large one or if it may contain <br>
non-english characters because vowel accents may cause the data to be rendered incorrectly. <br>

## Other pandas commands to manipulate the data:
## df[ ] , .loc , .iloc, .at and an introduction to indexing

In [17]:
#this syntax is wrong : df['tip','day']
#this syntax [[double square brackets]] is correct
df[['tip', 'day']]
#the df command produces an indexed data series of just two data columns that were passed into the argument.

tip   day
0    1.01   Sun
1    1.66   Sun
2    3.50   Sun
3    3.31   Sun
4    3.61   Sun
5    4.71   Sun
6    2.00   Sun
7    3.12   Sun
8    1.96   Sun
9    3.23   Sun
10   1.71   Sun
11   5.00   Sun
12   1.57   Sun
13   3.00   Sun
14   3.02   Sun
15   3.92   Sun
16   1.67   Sun
17   3.71   Sun
18   3.50   Sun
19   3.35   Sat
20   4.08   Sat
21   2.75   Sat
22   2.23   Sat
23   7.58   Sat
24   3.18   Sat
25   2.34   Sat
26   2.00   Sat
27   2.00   Sat
28   4.30   Sat
29   3.00   Sat
..    ...   ...
214  6.50   Sat
215  1.10   Sat
216  3.00   Sat
217  1.50   Sat
218  1.44   Sat
219  3.09   Sat
220  2.20   Fri
221  3.48   Fri
222  1.92   Fri
223  3.00   Fri
224  1.58   Fri
225  2.50   Fri
226  2.00   Fri
227  3.00   Sat
228  2.72   Sat
229  2.88   Sat
230  2.00   Sat
231  3.00   Sat
232  3.39   Sat
233  1.47   Sat
234  3.00   Sat
235  1.25   Sat
236  1.00   Sat
237  1.17   Sat
238  4.67   Sat
239  5.92   Sat
240  2.00   Sat
241  2.00   Sat
242  1.75   Sat
243  3.00  Thur

[244 rows x 2 columns]

In [ ]:
#to demonstrate an important concept in pandas - location
#this command produces output from row 200 up to but not including 206
#pandas associates index with the dataframe and that the index is a part of it
#this command relates to the position of the rows by default and does not take the index into account
df[0:5]

In [ ]:
#by specifying ".loc" in this version a different output is received because it is linked directly to the index labelling 
#instead.
df.loc[0:5]

In [ ]:
#by inserting arguments, a sub-selection of data-points in rows 200-205 can be output
df[['tip', 'sex']][200:206]

In [ ]:
#a different syntax is required when using .loc
df.loc[200:206, ['tip' , 'sex']]

In [ ]:
# .loc and .iloc identify the same data point when an index it entered as an argument
df.iloc[204]

In [ ]:
df.loc[204]

In [ ]:
#.at used to access a single value for a row/column label pair.
df.at[204, 'day']

In [ ]:
#this command includes a value at either side of a colon to indicate particular records
#Using the == symbol, the output will be in Boolean format True/False (just one = will not produce output).
df.loc[200:204, 'day'] == 'Sun'

In [ ]:
#this command does not include any values on either side of the colon which indicate all rows are included
df.loc[:, 'day'] == 'Sun'

In [ ]:
#this is a useful tool using Boolean values to select item
#the command outputs only the results== Sun from the set in the previous cell
df.loc[df.loc[:, 'day'] == 'Sun']

In [ ]:
#rename the above output as x
x=df.loc[df.loc[:, 'day'] == 'Sun']

In [ ]:
#call for the 190th row
#using loc,linked to original index values
x.loc [190]

In [ ]:
#to call for what is the 190th row in the above cell use 75 instead
#this time using iloc,linked to new position in a dataframe with 76 rows (0-75 in computer counting)
x.iloc [75]

## Plots to describe the tips dataset

Data visualization refers to taking the data and converting it into a more visual form. <br>
Instead of viewing numbers on spreadsheets, a picture is produced that represents that information.<br>
This is a useful tool for optimal communication of ideas that are backed up with statistics. <br>
It is worth bearing in mind that there is no "best" way to plot data.  It is best to understand your audience in order to choose the most useful kind of visualisation that will address their interests.<br>

In [ ]:
import seaborn as sns
#Plot the iris data set with a pairplot
sns.pairplot(df, hue = "class")

In [ ]:
import seaborn as sns
#Plot the tips data set with a pairplot
df=sns.load_dataset("tips")
sns.pairplot(df, hue = "tip")

**Summary of Section 1 Descriptive Statistics and Plots**

In this section, I have demonstrated a selection of  pandas and numpy methods that can be used to quickly produce descriptive statistics relating to a dataset.
* tips.info()
* tips.shape
* tips.head()
* tips.tail (3)
* tips.describe()
* df[ ] , .loc , .iloc, .at and an <br>
I also briefly discussed indexing.

Following on, I gave XX examples using matplotlib and seaborn of introductory plots.

# 2. To the above jupyter notebook add a section that discusses and analyses whether there is a relationship between the total bill and tip amount.

In my research for this project, I have carried out many google searches.<br>
This has yielded results of many different coding scripts which produce plots but very few offer any insights as <br>
what can be learned from the dataset.
On his [website](https://amitkushwaha.co.in/data-visualization-part-1.html) Amit Kushwaha provides a tutorial in Data Visualisation using seaborn that references the tips dataset.

In [ ]:
print("Data Visualisation - relationship between total bill and tip value")
#code verbatim fromhttps://amitkushwaha.co.in/data-visualization-part-1.html
tips = sns.load_dataset("tips")
#I know from earlier exploration that there are 7 different categories of data: total_bill,tip,sex,smoker,day,time,size
#which can be selected and displayed on axes as required
#select total_bill and tip to compare
sns.relplot(x="total_bill", y="tip", data=tips);

#Breakdown of the script:
#relplot is short relationship plot, a function that enables plotting on data-aware grids
#The scatterplot() is the default kind in relplot() (it can also be forced by setting kind="scatter")
#the script identifies the variables to display on x and y axis and where the data originates from

**Comment:** 
The plot above compares tip with total bill.  The visualisation confirms what was set out in an earlier command appearing in this notebook (tips.describe()) i.e.,
the datapoints centre around the mean values of just less than \\$20 and \\$3 tip (roughly 15% tip)
There is no **significant** deviation from the standard tip value of 10-16%

In [ ]:
#code verbatim fromhttps://amitkushwaha.co.in/data-visualization-part-1.html
print("Data Visualisation - relationship between total bill, tip value and smoker status")
sns.relplot(x="total_bill", y="tip", hue="smoker", data=tips);

#This script revisits the bill versus tip plot, but adds a third dimension with the inclusion of smoker status
#the "hue" parameter lets seaborn know the different data points in a particular grouping should be distinguished by colour.

In [ ]:
#Plot the relation graph between total bill and tip
#code amended from https://amitkushwaha.co.in/data-visualization-part-1.html
sns.relplot('total_bill', 'tip', data=tips, kind='line')


**Comment:** 
I have included this plot to demonstrate how certain renderings are not suitable<br>
for particular data analyses. The above plot with its erratic movement is difficult to understand.  

In [ ]:
#Categorical scatterplots

#code amended from https://amitkushwaha.co.in/data-visualization-part-1.html

sns.catplot(x="total_bill", y="tip", data=tips)

**Comment:** 
This is another example of a plot that is somewhat unsuited to the data.
A catplot (categorial scatterplot) shows the frequencies of categories. There are two many categories of tips (227)<br>
to display clearly.
If the categories on each axis are switched, the result is much the same.

***

Ranjeet Jain has a well laid out and easy to navigate website on kaggle.com. <br>
In addition to clear technical instructions, he offers some interpretation of the plots created as well as a comprehensive<br>
overview of the source and importance of the tips dataset.
According to Ranjeet:
>Restaurant managers need to know which factors matter when they assign tables to food servers. For the sake of staff morale, 
>they usually want to avoid either the substance or the appearance of 
>unfair treatment of the servers, for whom tips (at least in restaurants in the United States) 
>are a major component of pay.

In [ ]:
#code verbatim from https://www.kaggle.com/ranjeetjain3/different-types-of-charts
sns.jointplot(x = 'total_bill', y = 'tip', data = tips)

**Comment:**
Jointplot produces a plot of two variables,showing the relationship with a bivariate plot.
It includes the individual distributions represented in histogram format to the sides.
Ranjeet Jain summarises that this plot illustrates that tips are predominantly in the range from \\$1 - \\$5.
This can be viewed from the concentration of dots in the bottom left hand corner of the plot.

In [ ]:
#code verbatim from https://www.kaggle.com/ranjeetjain3/different-types-of-charts
sns.jointplot(x = 'total_bill', y = 'tip', data = tips ,kind = 'hex')

**Comment:**
Ranjeet Jain describes this as "another clean visualization of viewing the graph" where the darker region indicates a greater
number of data points present.  
Personally, I find this hex representation easier to interpret than a scatterplot. 

In [ ]:
#code verbatim from https://www.kaggle.com/ranjeetjain3/different-types-of-charts
sns.jointplot(x = 'total_bill', y = 'tip', data = tips ,kind = 'reg')

**Comment:**
The inclusion of the code: kind = 'reg' incorporates a simple regression line which Ranjeet Jain states shows that the tip value increases based on the total bill.
This corresponds with my finding in the first plot that a tip usually represents 15% of the total bill, therefore the tip value will increase at a similar rate to the total bill. 
Regression is a key concept in data analysis. In this plot, the inclusion of a regression line, enables better understanding of the data and whether there may be a linear relationship between them.
The regression line on the bivariate part of the plot indicates that there is a normal linear regression in this case. 
The regression line on the marginal areas shows distribution instead and both can be classified as having normal distribution as indicated by the customary bell-shaped curve.

In [ ]:
#https://www.kaggle.com/ranjeetjain3/different-types-of-charts
sns.jointplot(x = 'total_bill', y = 'tip', data = tips ,kind = 'kde')
#another way of visualization. This graph shows that darker the region more number of tips are given at that region

**Comment:**
Jointplots can also be used to display Kernel Density Function (selected by using kind = 'kde') which plots the density of observations on one axis with height along the other axis.
Similar to the hex representation density is indicated by a darker colour.

**Summary of Section 2 - Is there a relationship between the total bill and tip amount?**

In this section I explored the relationship between total bill and tip amount while also going into detail about Linear Regression - a key concept in Data Analysis using plots.
The relationship was proven through the following
*
*
*
I referenced websites created by XXYYZZ who have studied the dataset and arrived at similar conclusions to me.


# 3. Using the same notebook, analyse the relationship between the variables within the dataset. You are free to interpret this as you wish — for example, you may analyse all pairs of variables, or select a subset and analyse those.


In [ ]:
#Plot the relation graph between total bill and tip
#code verbatim from https://amitkushwaha.co.in/data-visualization-part-1.html
#sns.relplot('total_bill', 'tip', data=tips, kind='line')
sns.relplot('day', 'tip', data=tips, kind='line')

In [ ]:
**GENDER** see https://dfrieds.com/data-analysis/groupby-python-pandas
    that bloke thinks male means server was male.
    How can I prove otherwise...using data?
    then I can question some of his findings.

In [ ]:
   https://seaborn.pydata.org/examples/grouped_boxplot.html
        #verbatim from https://www.kaggle.com/ranjeetjain3/different-types-of-charts

#plt.show
#sns.boxplot(x = 'day', y= 'total_bill', data = tips)

**Comment:**
A box plot  shows the distribution of quantitative data in a way that facilitates comparisons 
between variables or across levels of a categorical variable. The box shows the quartiles of the dataset while the 
whiskers extend to show the rest of the distribution, except for points that are determined to be “outliers” using a method 
that is a function of the inter-quartile range. 
FROMMYGITHUB
Box plots (or box-and-whisker plot)get the name from the box in the middle of the diagram and are a way to display numerical data via their quartiles. In the figure that is rendered:

the minimum is represented by the lowermost line (a "whisker")
the maximum value is the uppermost line (a "whisker")
the lowermost end of each box is quartile 1, the uppermost represents quartile 3
the second quartile (the median) is the line inside the box.
the circles that appear on the plot for "sepal width in cm" indicate data that is outside the median, unusual recordings compared to the majority of the data
a compact box indicates less variation in the values, as shown in the box plot directly below relating to iris sepals while the petal length can vary from approx 1,5 cm to 5 cm. I have written scripts that display univariate (single variable) and multivariate (two or more variables) versions of boxplots

# Conclusion

***

### _Seaborn_
In order to best utilise the functionality of seaborn, the dataset should be in "long form" or "tidy" format:
Each variable is a column
Each observation is a row
Therefore an effort should be made to collect and arrange the data this way. 

### _Seaborn V pandas_
  



In [ ]:
https://seaborn.pydata.org/examples/faceted_histogram.html
 
        https://seaborn.pydata.org/examples/grouped_violinplots.html
            https://seaborn.pydata.org/examples/regression_marginals.html

In [ ]:
Summary
In this notebook I
1. bla
2.bla
3.bla
4. bla
